src/main/cpp/js.cpp

```diff
@@ -0,0 +1,1071 @@
+
+    case u8'-':
+        return str.starts_with(u8"--") ? decrement : str.starts_with(u8"-=") ? minus_equal : minus;
+
+    case u8'.': return str.starts_with(u8"...") ? ellipsis : dot;
+
+    case u8'/':
+        if (is_jsx) {
+            return str.starts_with(u8"/>") ? jsx_tag_self_close : divide;
+        }
+        else {
+            return str.starts_with(u8"/=") ? divide_equal : divide;
+        }
+
+    case u8':': return colon;
+    case u8';': return semicolon;
+
+    case u8'<': {
+        if (is_jsx) {
+            return str.starts_with(u8"</>") ? jsx_fragment_close
+                : str.starts_with(u8"<>")   ? jsx_fragment_open
+                : str.starts_with(u8"</")   ? jsx_tag_end_open
+                                            : jsx_tag_open;
+        }
+        return str.starts_with(u8"<<=") ? left_shift_equal
+            : str.starts_with(u8"<<")   ? left_shift
+            : str.starts_with(u8"<=")   ? less_equal
+                                        : less_than;
+    }
+
+    case u8'=': {
+        if (is_jsx) {
+            return jsx_attr_equals;
+        }
+        return str.starts_with(u8"===") ? strict_equals
+            : str.starts_with(u8"==")   ? equals
+            : str.starts_with(u8"=>")   ? arrow
+                                        : assignment;
+    }
+
+    case u8'>': {
+        if (is_jsx) {
+            return jsx_tag_close;
+        }
+        return str.starts_with(u8">>>=") ? unsigned_right_shift_equal
+            : str.starts_with(u8">>>")   ? unsigned_right_shift
+            : str.starts_with(u8">>=")   ? right_shift_equal
+            : str.starts_with(u8">>")    ? right_shift
+            : str.starts_with(u8">=")    ? greater_equal
+                                         : greater_than;
+    }
+
+    case u8'?':
+        return str.starts_with(u8"??=") ? nullish_coalescing_equal
+            : str.starts_with(u8"??")   ? nullish_coalescing
+            : str.starts_with(u8"?.")   ? optional_chaining
+                                        : conditional;
+
+    case u8'[': return left_bracket;
+    case u8']': return right_bracket;
+
+    case u8'^': return str.starts_with(u8"^=") ? bitwise_xor_equal : bitwise_xor;
+
+    case u8'{': return left_brace;
+
+    case u8'|':
+        return str.starts_with(u8"||=") ? logical_or_equal
+            : str.starts_with(u8"||")   ? logical_or
+            : str.starts_with(u8"|=")   ? bitwise_or_equal
+                                        : bitwise_or;
+
+    case u8'}': return right_brace;
+
+    case u8'~': return bitwise_not;
+
+    default: return {};
+    }
+}
+
+namespace {
+/// @brief  Common JS and JSX highlighter implementation.
+bool highlight_javascript_impl(
+    Non_Owning_Buffer<Token>& out,
+    std::u8string_view source,
+    std::size_t start_index,
+    std::size_t length,
+    bool initial_can_be_regex,
+    bool is_at_start_of_file,
+    Jsx_State jsx_state,
+    const Highlight_Options& options
```

Comment:
Can you please make a `struct Highlighter` instead, similar to how the HTML and CSS highlighters do it?

This avoids creating such functions with tons of parameters. It looks like `Jsx_State` is value-initialized every time you recurse anyway, so that could just be a class member with default construction.